### PR TITLE
feat(scripts): add swarm-model config CLI (supersedes #1857)

### PR DESCRIPTION
**Pull request by zaxbysauce**
The design and both implementations (Node + PowerShell) are @goodie1972's work from #1857. This PR carries their original commit as its base and adds review fixes on top. It's a separate PR only because their branch lives on a fork this session can't push to; #1857 is closed in favor of this one.

## Summary

Adds an interactive helper CLI — `scripts/swarm-model/` (Node.js, cross-platform) and `scripts/swarm-model.ps1` (PowerShell, Windows) — that reads available providers/models from `opencode.json` and rewrites an agent's `model`/`temperature` in `opencode-swarm.json`, backing up the file before every write.

Imported from #1857 and hardened against 17 validated review findings:

- **F-01 (was CRITICAL):** interactive mode crashed on every run — `ask` was called but never imported. Now exported from `ui.js` and imported in `interactive.js`.
- **F-07:** `list`/`help` hung on a TTY because readline was created at import; now created lazily.
- **F-06:** agent entries with no `model` field no longer crash agent selection.
- **F-04:** the Node picker now offers a quit option at every selection step.
- **F-02:** backup filenames use a full `YYYYMMDDHHMMSS` (second-resolution) timestamp, so writes in the same 10-minute window no longer overwrite each other.
- **F-03:** PowerShell `ConvertTo-Json -Depth 10` — nested agent fields (`fallback_models`, `reasoning`, `thinking`) are no longer truncated on write.
- **F-08/F-09/F-13:** graceful malformed-JSON handling, a clear error on a missing `--swarm-config`/`--opencode-config` value, and robust `os.homedir()` resolution (no import-time crash when `HOME`/`USERPROFILE` are unset).
- **F-05/F-11/F-15/F-16/F-17:** consistent model splitting, removed dead `main`/duplicated constants, cleaner unset-temperature display, and README fixes (example/table consistency, PS flags, i18n note, backup format).
- **F-12:** added a `node:test` suite and wired it into CI so the utility's regression tests are actually gated.

This is a standalone contributor utility under `scripts/`; it is not part of the published `opencode-swarm` package.

## Invariant audit
- 1 (plugin init):       not touched — no changes under `src/`; only `scripts/` utility + one CI step.
- 2 (runtime portability): not touched — plugin shape (`src/`) unchanged; the utility is a standalone Node CLI that uses `os.homedir()` for cross-platform home resolution.
- 3 (subprocesses):       not touched (plugin) — the new `test/cli.test.js` spawns short-lived `node` children for CLI integration; each exits within the test and is not a plugin-runtime subprocess.
- 4 (.swarm containment): not touched — the tool writes only to user-supplied config paths; tests use `fs.mkdtempSync(os.tmpdir())`; no `.swarm/` writes.
- 5 (plan durability):    not touched.
- 6 (test_runner safety): not touched — utility tests run via `node --test` in a dedicated `quality`-job step, not the plugin's `test_runner`.
- 7 (test writing):       touched (adjacent) — added Node built-in `node:test` (not `bun:test`) for this standalone utility, outside `src/tests`; tests are deterministic and isolated (per-test `mkdtemp`, injected IO reset in `finally`). Evidence: `cd scripts/swarm-model && node --test` → 17/17 pass; each fix proven load-bearing by reverting it and observing the matching test fail.
- 8 (session state):      not touched.
- 9 (guardrails/retry):   not touched.
- 10 (chat/system msg):   not touched.
- 11 (tool registration): not touched — no plugin tools/agents added. A CI workflow step was added; evidence: `ci.yml` parses as valid YAML and its `if: needs.detect-release.outputs.is-release != 'true'` mirrors 15 existing sibling steps in the same `quality` job.
- 12 (release/cache):     touched — added `docs/releases/pending/1857-swarm-model-cli.md`. Evidence: satisfies the `pr-standards` scripts/-change release-fragment gate (an ADDED `docs/releases/pending/*.md`).

## Test plan
- `cd scripts/swarm-model && node --test` → **17/17 pass** (config/backup/provider-merge units + end-to-end interactive drive + CLI integration).
- Reversion proof: reverting F-01/F-07/F-13 makes the corresponding tests fail (`ReferenceError` / hang-timeout / import-time `TypeError`); the config/backup tests fail under the old `slice(0,14)` timestamp logic.
- Runtime probes: `help`/`list` exit 0 with stdin held open (were hanging); `list --swarm-config` with no value exits 1 with a clear message; `env -u HOME -u USERPROFILE ... help` exits 0.
- Repo gates: `bun run typecheck` clean, `bun run lint` (biome) clean, `bun run build` green, `git diff --check` clean. (`scripts/**/*.js` is outside biome/tsc scope by repo convention; covered by `node --check` + the new CI test step.)

### CODE REVIEW

File: scripts/swarm-model/src/config.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    let dest = backupPath;
+    if (fs.existsSync(backupPath)) {
+      // Full YYYYMMDDHHMMSS second-resolution stamp (matches README + the
+      // PowerShell impl). The earlier slice(0,14) of the raw ISO string kept
+      // the `-`/`T` separators and truncated to ~10-minute resolution, so
+      // backups within the same 10-minute window collided and overwrote each
+      // other; stripping all non-digits first avoids that.
+      const ts = new Date().toISOString().replace(/\D/g, '').slice(0, 14);
+      dest = filePath + '.' + ts + '.bak';
+    }
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply scripts/swarm-model/src/config.js:40]: Fixed in afa9ba9. The timestamped backup now gets a `-N` suffix when the slot's already taken, so a second write in the same second can't overwrite an earlier backup. Mirrored the same loop in the PowerShell version, and added a test that asserts existing backups are never clobbered.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/swarm-model/src/cli.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  let swarmConfig = DEFAULT_SWARM_CONFIG;
+  let openCodeConfig = DEFAULT_OPENCODE_CONFIG;
+
+  for (let i = 0; i < args.length; i++) {
+    if (args[i] === '--swarm-config' || args[i] === '--opencode-config') {
+      const flag = args[i];
+      const value = args[i + 1];
+      if (!value) {
+        throw new Error(`Missing value for ${flag}`);
+      }
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply scripts/swarm-model/src/cli.js]: Fixed in afa9ba9. A following token that starts with `--` is now treated as a missing value, so `--swarm-config --opencode-config x` errors instead of consuming the next flag as the path. Added a test covering it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by zaxbysauce** [comment]
## 🤖 Multi-Stage PR Review
**Pipeline:** Esper3.1 (Qwen3.6-35B-A3B) (context pack) → Mellum2-Claude (bug/logic lens) + Qwythos-9B-v2 (sec/perf lens) (parallel explore, distinct lenses) → Qwythos-9B-v2 (critiquing Mellum findings) ↔ Mellum2-Claude (critiquing Qwythos findings) (cross-critique) → Esper3.1 (Arbiter) (arbiter: blind-spot + synthesize)
**Commit reviewed:** `afa9ba908a8f`

---

### 🔍 PR Intent

> Reconstructed obligation list (from PR text, issue, commits, changed tests, changed docs, changed interfaces).

- **O-001**: Add a standalone Node.js CLI tool under `scripts/swarm-model/` for interactive configuration of swarm agent models and temperatures.
- **O-002**: Add a PowerShell equivalent for Windows.
- **O-003**: Ensure the tool gracefully handles malformed JSON and missing config files without crashing.
- **O-004**: Implement a robust backup mechanism with unique filenames (timestamps) to prevent data loss.
- **O-005**: Add a `node:test` suite and integrate it into the CI workflow.
- **O-006**: Ensure the CLI does not hang on non-interactive terminals (TTY handling).
- **O-007**: Provide clear error messages for invalid command-line arguments.

### 📦 Implementation Summary

> What the code actually does, summarised independently of the obligation list.

The PR adds a new CLI tool (`scripts/swarm-model/`) with a Node.js implementation and a PowerShell script. The Node.js version uses `readline` lazily to avoid hanging on non-interactive terminals. It reads agent configurations from `opencode-swarm.json` and available models from `opencode.json`. The tool supports listing agents, showing help, and an interactive mode to change models and temperatures. It includes a backup mechanism that appends timestamps to backup files to ensure uniqueness. A `node:test` suite is added, and a CI step is introduced to run these tests. The PowerShell version mirrors the Node.js functionality but with Chinese prompts and PowerShell-specific features.

### ✅ / ⚠️ / ❌ Intended vs Actual

| Obligation | Status | Evidence (file:line) |
|---|---|---|
| O-001 | `SUPPORTED` | `scripts/swarm-model/src/cli.js` - CLI entry point |
| O-002 | `SUPPORTED` | `scripts/swarm-model.ps1` - PowerShell implementation |
| O-003 | `PARTIALLY_SUPPORTED` | `scripts/swarm-model/src/config.js:14` - Swarm config throws, opencode config does not |
| O-004 | `SUPPORTED` | `scripts/swarm-model/src/config.js:27` - Timestamped backups |
| O-005 | `SUPPORTED` | `scripts/swarm-model/test/` - Tests and `ci.yml` integration |
| O-006 | `SUPPORTED` | `scripts/swarm-model/src/ui.js:9` - Lazy readline |
| O-007 | `SUPPORTED` | `scripts/swarm-model/src/cli.js:21` - Argument validation |

### 🚨 Confirmed Findings

#### [HIGH] Asymmetric error handling for config parsing
- **Location**: `scripts/swarm-model/src/config.js:14` vs `scripts/swarm-model/src/config.js:72`
- **Why it matters**: The tool crashes on malformed swarm config (`readSwarmConfig`) but gracefully handles malformed opencode config (`getOpenCodeProviders`). This inconsistency could lead to unexpected failures for users with corrupted swarm config files, which is a more critical failure mode than missing providers.
- **Evidence**: `readSwarmConfig` throws an `Error` on parse failure. `getOpenCodeProviders` catches the error and returns an empty object. The PR description claims "graceful malformed-JSON handling" but only applies it to one config source.
- **Fix direction**: Make `readSwarmConfig` handle parse errors gracefully (e.g., log a warning and exit) or document the crash behavior clearly. Given the tool is a contributor utility, crashing is less severe than for a runtime plugin, but consistency is key.

#### [MEDIUM] No argument validation for config paths
- **Location**: `scripts/swarm-model/src/cli.js:21`
- **Why it matters**: The CLI does not validate if the provided config paths exist before attempting to read them. Errors occur deep in the flow, making debugging harder for users.
- **Evidence**: `parseArgs` only checks for the presence of a value, not its existence. The existence check is in `readSwarmConfig` and `getOpenCodeProviders`.
- **Fix direction**: Add a simple existence check in `parseArgs` or at the start of `main` to provide immediate feedback.

#### [LOW] PowerShell version lacks input validation for temperature
- **Location**: `scripts/swarm-model.ps1:342`
- **Why it matters**: The PowerShell version allows entering a temperature value that is not a number, which could lead to unexpected behavior or errors when the value is used.
- **Evidence**: The `Read-Host` input is not validated as a number before being used. The Node.js version has `parseFloat` and range checks.
- **Fix direction**: Add input validation in the PowerShell version to match the Node.js version.

### 🔬 Unverified but Plausible Risks

- **Risk**: The lazy readline initialization might not handle all edge cases for non-TTY environments.
  - *Why suspicious*: The test `cli.test.js:62` checks for hanging but might not cover all CI environments.
  - *What would verify it*: Run the tool in various CI environments and non-interactive scripts.

### 🧪 Test / Coverage Gaps

- **Gap**: No tests for the PowerShell version.
  - *Evidence*: Tests are only in `scripts/swarm-model/test/` and cover the Node.js version.

### 📋 Shipped-vs-Claimed Gaps

- **Gap**: The PR description claims "graceful malformed-JSON handling" but it is only implemented for one config source.
  - *Evidence*: `readSwarmConfig` throws, `getOpenCodeProviders` does not.

### 📝 Merge Recommendation

**APPROVE_WITH_FIXES**

The PR is well-structured and adds significant value, but the asymmetric error handling and lack of input validation in the PowerShell version are real issues that should be fixed before merging.

| Check | Result |
|---|---|
| No CRITICAL findings | ✅ |
| No unresolved STEALTH_CHANGE | ✅ |
| No UNSUPPORTED obligations | ✅ |
| Test coverage adequate | ✅ |
| No hardcoded secrets | ✅ |
| All async errors handled | ✅ |
| Input validation present | ⚠️ |
| No broken agent role boundaries | ✅ |
| Prompt format contracts intact | ✅ |
| Lockfile consistent | ✅ |

---
<sub>🔒 Reviewed by a 3-model cross-family adversarial debate (architect → dual-lens parallel explorers → cross-critique → arbiter) for high recall with low false-positive noise. Findings are advisory — verify before acting.</sub>

**Comment by zaxbysauce** [comment]
Went through the three findings against the source. They're advisory and I don't think any needs a code change — reasoning below.

**Asymmetric config error handling — intended.** `readSwarmConfig` handles the config the tool *writes to*, so on malformed JSON it fails with a clear message and exit 1 (`Failed to parse swarm config <path>: ...`) instead of returning `{}`. Returning empty there would treat a corrupted-but-real config as empty and clobber it on the next write. `getOpenCodeProviders` is read-only provider discovery, so degrading to `{}` is safe. Both are handled; fail-loud on the file we write, degrade on the file we only read.

**Path-existence check — would break auto-create.** A missing `--swarm-config` path isn't an error: `readSwarmConfig` creates it (the `!fs.existsSync` branch → `mkdirSync` + write + `Created:`), which is one of the tool's features. A missing `--opencode-config` returns `{}` (no discovered providers). An up-front "must exist" check would break both.

**PowerShell temperature validation — already present.** `swarm-model.ps1:331` does `[double]::TryParse($tempInput.Trim(), [ref]$tempNum)` with a 0.0–2.0 range check and keeps the current value on invalid input, matching the Node `parseFloat` path.

Fair call on the PowerShell test gap — there's no `pwsh` in CI, so that side stays manually verified.

---
_Generated by [Claude Code](https://claude.ai/code)_
